### PR TITLE
Add a badlist to mempool

### PR DESCRIPTION
**Pull request by gregorycollins**
Currently we don't remember when transactions fail pre-block validation. Since pre-block validation only happens when we call `mempoolGetBlock`, and nodes pick a chain to mine on at random, there is always a window where:

  - we call `mempoolGetBlock`, the transaction gets dropped
  - we initiate a connection with a peer that still has the bad tx
  - the bad tx gets regossiped back to us. Since we don't keep a badlist, we accept it, and happily re-gossip it to other peers.

Lars thinks that a precise mapping of bad transactions (keeping the hashes until they TTL-expire) might be memory-intensive, but we should implement this fix for now and follow up with (perhaps) a probabilistic check later.

### CODE REVIEW

File: src/Chainweb/Mempool/InMem.hs [hunk 1]
```diff
@@ -260,16 +267,28 @@ getBlockInMem :: forall t .
 getBlockInMem cfg lock txValidate bheight phash size0 = do
     withMVar lock $ \mdata -> do
         !psq0 <- readIORef $ _inmemPending mdata
+        -- TODO: temporary: delete this in preference for Emmanuel's TTL branch
```
Comment (fosskers) [inline src/Chainweb/Mempool/InMem.hs]: Can this be dealt with yet?

File: src/Chainweb/Mempool/InMem.hs [hunk 2]
```diff
@@ -260,16 +267,28 @@ getBlockInMem :: forall t .
...
+        -- which passes block creation time as a parameter
+        now <- getCurrentTimeIntegral
+        !badmap <- pruneBadMap now <$!> readIORef (_inmemBadMap mdata)
         let !psq = HashMap.map decodeTx psq0
-        (!psq', out) <- go psq size0 []
-        let ins !h !t = HashMap.insert (hasher t) t h
-        -- put the pending txs back into the map.
+        (T3 !psq' !badmap' out) <- go psq badmap size0 []
```
Comment (fosskers) [inline src/Chainweb/Mempool/InMem.hs]: They probably don't need to be banged if already in a `T3`.

File: src/Chainweb/Mempool/InMem.hs [hunk 3]
```diff
@@ -340,17 +363,18 @@ getBlockInMem cfg lock txValidate bheight phash size0 = do
...
-    go !psq !remainingGas !soFar = do
+       -> IO (T3 (HashMap TransactionHash t) BadMap (Vector t))
+    go !psq !badmap !remainingGas !soFar = do
         nb <- nextBatch psq remainingGas
         if null nb
-          then return (psq, V.concat soFar)
+          then return $! T3 psq badmap (V.concat soFar)
           else do
-            (good, psq') <- validateBatch psq $! V.fromList nb
+            (T3 good psq' badmap') <- validateBatch psq badmap $! V.fromList nb
```
Comment (fosskers) [inline src/Chainweb/Mempool/InMem.hs]: Don't need parens here.

File: src/Chainweb/Mempool/InMem.hs (hunk 1 above)
Comment (gregorycollins) [reply src/Chainweb/Mempool/InMem.hs]: no

Comment (gregorycollins) [reply src/Chainweb/Mempool/InMem.hs]: ...I expect this one to get in first at this rate

**Comment by larskuhtz** [comment]
This has been tested on devnet over the past > 12h.

**Comment by mightybyte** [comment]
I was going to try to merge this tonight but wasn't quite sure how to resolve the conflict.

File: src/Chainweb/Mempool/InMem.hs (hunk 1 above)
Comment (gregorycollins) [reply src/Chainweb/Mempool/InMem.hs]: ... the time check is being done elsewhere and we are only using 'now' to clean out badmap -- i think this is ok

**Pull request merged**